File: core/control-api/internal/httpapi/skills.go
// SPDX-License-Identifier: Apache-2.0

package httpapi

import (
	"errors"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/batonos/baton/core/control-api/internal/channel"
	"github.com/batonos/baton/core/control-api/internal/skills"
	"github.com/batonos/baton/core/pkg/spi/audit"
	"github.com/batonos/baton/core/pkg/spi/console"
	spi "github.com/batonos/baton/core/pkg/spi/store"
)


const (
	CodeSkillDigestMismatch = "SKILL_DIGEST_MISMATCH"
	CodeSkillFetchFailed    = "SKILL_FETCH_FAILED"
	CodeSkillTargetInvalid  = "SKILL_TARGET_INVALID"





	CodeSkillNoMountPath = "SKILL_NO_MOUNT_PATH"
)

type skillView struct {
	SkillID string `json:"skill_id"`
	Name    string `json:"name"`
	Version string `json:"version"`



	SHA256    string `json:"sha256"`
	SourceURL string `json:"source_url"`
	SizeBytes int64  `json:"size_bytes"`
	CreatedAt string `json:"created_at"`




	Cached bool `json:"cached"`

	Installs []skillInstallView `json:"installs"`
}

type skillInstallView struct {
	TargetKind string `json:"target_kind"`
	Target     string `json:"target"`
	CreatedAt  string `json:"created_at"`
}

func (a *API) toSkillView(s spi.Skill, installs []spi.SkillInstall) skillView {
	v := skillView{
		SkillID:   s.SkillID,
		Name:      s.Name,
		Version:   s.Version,
		SHA256:    s.SHA256,
		SourceURL: s.SourceURL,
		SizeBytes: s.SizeBytes,
		CreatedAt: s.CreatedAt.UTC().Format(time.RFC3339),
		Cached:    a.SkillCache != nil && a.SkillCache.Has(s.SHA256),
		Installs:  []skillInstallView{},
	}
	for _, in := range installs {
		v.Installs = append(v.Installs, skillInstallView{
			TargetKind: in.TargetKind,
			Target:     in.Target,
			CreatedAt:  in.CreatedAt.UTC().Format(time.RFC3339),
		})
	}
	return v
}







func (a *API) handleSkillCreate(w http.ResponseWriter, r *http.Request) {
	p, ok := a.requireRole(w, r, console.RoleOperator)
	if !ok {
		return
	}
	var body struct {
		Name      string `json:"name"`
		Version   string `json:"version"`
		SourceURL string `json:"source_url"`
		SHA256    string `json:"sha256"`
	}
	if !a.decodeJSON(w, r, &body) {
		return
	}
	body.Name = strings.TrimSpace(body.Name)
	if body.Name == "" || body.SourceURL == "" || body.SHA256 == "" {
		a.fail(w, r, http.StatusBadRequest, CodeInvalidRequest,
			"a skill needs a name, a source_url and a sha256",
			"The digest is not optional: it is the only check on what the source serves.", nil)
		return
	}
	if body.Version == "" {
		body.Version = "0.0.0"
	}

	size, err := a.SkillCache.Fetch(r.Context(), body.SourceURL, body.SHA256)
	switch {
	case errors.Is(err, skills.ErrDigestMismatch):
		a.fail(w, r, http.StatusBadRequest, CodeSkillDigestMismatch,
			"the source served bytes that do not match the declared digest",
			"Either the download was corrupted or the source is serving something else. "+
				"Check the digest against the directory before retrying.",
			map[string]any{"source_url": body.SourceURL, "declared": body.SHA256})
		return
	case errors.Is(err, skills.ErrDigestMalformed):
		a.fail(w, r, http.StatusBadRequest, CodeInvalidRequest,
			"the digest is not sha256:<64 hex characters>",
			"Copy it from the directory listing exactly, including the sha256: prefix.", nil)
		return
	case err != nil:
		a.fail(w, r, http.StatusBadGateway, CodeSkillFetchFailed,
			"the archive could not be fetched",
			"The control plane fetches and checks skills so nodes do not have to: a node downloads them only "+
				"from the control plane, never from the source. "+
				"Check that this host can reach the source.",
			map[string]any{"source_url": body.SourceURL, "error": err.Error()})
		return
	}

	created, err := a.Store.Skills().Create(r.Context(), spi.Skill{
		TenantID:  p.Identity.TenantID,
		Name:      body.Name,
		Version:   body.Version,
		SHA256:    body.SHA256,
		SourceURL: body.SourceURL,
		SizeBytes: size,
	})
	if err != nil {
		a.failStore(w, r, err)
		return
	}
	a.auditSkill(r, p.Subject, "skill.created", "create", created.SkillID)
	a.writeJSON(w, http.StatusCreated, a.toSkillView(*created, nil))
}

func (a *API) handleSkillList(w http.ResponseWriter, r *http.Request) {
	p, ok := a.requireRole(w, r, console.RoleViewer)
	if !ok {
		return
	}
	list, err := a.Store.Skills().List(r.Context(), p.Identity.TenantID)
	if err != nil {
		a.failStore(w, r, err)
		return
	}
	out := make([]skillView, 0, len(list))
	for _, s := range list {
		installs, err := a.Store.Skills().Installs(r.Context(), p.Identity.TenantID, s.SkillID)
		if err != nil {
			a.failStore(w, r, err)
			return
		}
		out = append(out, a.toSkillView(s, installs))
	}
	a.writeJSON(w, http.StatusOK, map[string]any{
		"apiVersion": "baton.mailloop.dev/v1alpha1",
		"kind":       "SkillList",
		"items":      out,



		"next_cursor": "",
	})
}

func (a *API) handleSkillDelete(w http.ResponseWriter, r *http.Request) {
	p, ok := a.requireRole(w, r, console.RoleOperator)
	if !ok {
		return
	}
	id := r.PathValue("skill_id")




	affected := a.nodesAffectedBySkill(r, p.Identity.TenantID, id)

	if err := a.Store.Skills().Delete(r.Context(), p.Identity.TenantID, id); err != nil {
		a.failStore(w, r, err)
		return
	}
	a.auditSkill(r, p.Subject, "skill.deleted", "delete", id)
	a.nudgeSkills(r, affected)
	w.WriteHeader(http.StatusNoContent)
}


func (a *API) handleSkillInstall(w http.ResponseWriter, r *http.Request) {
	p, ok := a.requireRole(w, r, console.RoleOperator)
	if !ok {
		return
	}
	id := r.PathValue("skill_id")
	var body struct {
		TargetKind string `json:"target_kind"`
		Target     string `json:"target"`
	}
	if !a.decodeJSON(w, r, &body) {
		return
	}
	if body.TargetKind != spi.SkillTargetNode && body.TargetKind != spi.SkillTargetTemplate {
		a.fail(w, r, http.StatusBadRequest, CodeSkillTargetInvalid,
			"target_kind must be node or template",
			"A node install names one workspace; a template install names a scene "+
				"and reaches nodes created later.", nil)
		return
	}
	if strings.TrimSpace(body.Target) == "" {
		a.fail(w, r, http.StatusBadRequest, CodeSkillTargetInvalid,
			"target is empty",
			"Name the node or the template. An empty template would collect every "+
				"node whose spec declared no scene.", nil)
		return
	}
	if _, err := a.Store.Skills().Get(r.Context(), p.Identity.TenantID, id); err != nil {
		a.failStore(w, r, err)
		return
	}















	if body.TargetKind == spi.SkillTargetNode {
		node, err := a.resolveNode(r, body.Target)
		if err != nil {
			a.failStore(w, r, err)
			return
		}
		body.Target = node.NodeID

























		if node.Runtime.Name != "" && node.Runtime.SkillsMountPath == "" {
			a.fail(w, r, http.StatusConflict, CodeSkillNoMountPath,
				"node "+node.DisplayName+" declares no adapter.skills.mountPath, so it has nowhere to put a skill",
				"Its Node Template has to say where skills go, for example:\n"+
					"  adapter:\n    skills:\n      mountPath: /workspace/.skills\n"+
					"BATON does not pick one, because another runtime reads another path. "+
					"Nothing was installed.", nil)
			return
		}
	}

	in, err := a.Store.Skills().Install(r.Context(), spi.SkillInstall{
		TenantID: p.Identity.TenantID, SkillID: id,
		TargetKind: body.TargetKind, Target: body.Target,
	})
	if err != nil {
		a.failStore(w, r, err)
		return
	}
	a.auditSkill(r, p.Subject, "skill.installed", "install", id+" -> "+in.TargetKind+":"+in.Target)
	a.nudgeSkills(r, a.nodesAffectedBySkill(r, p.Identity.TenantID, id))

	a.writeJSON(w, http.StatusCreated, skillInstallView{
		TargetKind: in.TargetKind,
		Target:     in.Target,
		CreatedAt:  in.CreatedAt.UTC().Format(time.RFC3339),
	})
}

func (a *API) handleSkillUninstall(w http.ResponseWriter, r *http.Request) {
	p, ok := a.requireRole(w, r, console.RoleOperator)
	if !ok {
		return
	}
	id := r.PathValue("skill_id")
	var body struct {
		TargetKind string `json:"target_kind"`
		Target     string `json:"target"`
	}
	if !a.decodeJSON(w, r, &body) {
		return
	}
	affected := a.nodesAffectedBySkill(r, p.Identity.TenantID, id)

	if err := a.Store.Skills().Uninstall(r.Context(), p.Identity.TenantID, id,
		body.TargetKind, body.Target); err != nil {
		a.failStore(w, r, err)
		return
	}
	a.auditSkill(r, p.Subject, "skill.uninstalled", "uninstall", id+" -> "+body.TargetKind+":"+body.Target)
	a.nudgeSkills(r, affected)
	w.WriteHeader(http.StatusNoContent)
}




func (a *API) auditSkill(r *http.Request, actor, event, action, target string) {
	if a.Log == nil {
		return
	}
	_, _ = a.Log.Append(r.Context(), audit.Record{
		Event: event, Actor: actor, ActorType: "user",
		Action: action, Target: target, Result: "ok",
		SourceIP: clientIP(r), RequestID: requestIDFrom(r), At: time.Now(),
	})
}








func (a *API) nodesAffectedBySkill(r *http.Request, tenantID, skillID string) []string {
	if a.Hub == nil {
		return nil
	}
	installs, err := a.Store.Skills().Installs(r.Context(), tenantID, skillID)
	if err != nil {
		return nil
	}
	nodes := map[string]bool{}
	for _, in := range installs {
		switch in.TargetKind {
		case spi.SkillTargetNode:
			nodes[in.Target] = true
		case spi.SkillTargetTemplate:
			list, _, err := a.Store.Nodes().List(r.Context(), spi.NodeFilter{TenantID: tenantID})
			if err != nil {
				continue
			}
			for _, n := range list {
				if n.Runtime.Template == in.Target {
					nodes[n.NodeID] = true
				}
			}
		}
	}
	out := make([]string, 0, len(nodes))
	for id := range nodes {
		out = append(out, id)
	}
	return out
}

func (a *API) nudgeSkills(r *http.Request, nodeIDs []string) {
	if a.Hub == nil {
		return
	}
	for _, id := range nodeIDs {
		conn, ok := a.Hub.Get(id)
		if !ok {
			continue
		}
		f, err := channel.NewFrame(channel.TypeChanged, "", channel.Changed{Kind: "skills"})
		if err != nil {
			continue
		}
		_ = conn.Send(r.Context(), f)
	}
}



type nodeSkillView struct {
	SkillID string `json:"skill_id"`
	Name    string `json:"name"`
	Version string `json:"version"`
	SHA256  string `json:"sha256"`



	Via string `json:"via"`















	CodeID string `json:"code_id,omitempty"`
}





func (a *API) handleAgentSkillList(w http.ResponseWriter, r *http.Request) {
	node, ok := a.requireNode(w, r)
	if !ok {
		return
	}
	list, err := a.Store.Skills().ForNode(r.Context(), node.TenantID, node.NodeID, node.Runtime.Template)
	if err != nil {
		a.failStore(w, r, err)
		return
	}






	codes, err := a.Store.Skills().CodesForNode(r.Context(), node.NodeID)
	if err != nil {



		a.Logger.Warn("skill codes for node", "node_id", node.NodeID, "error", err)
		codes = map[string]string{}
	}
	out := make([]nodeSkillView, 0, len(list))
	for _, s := range list {
		out = append(out, nodeSkillView{
			SkillID: s.SkillID, Name: s.Name, Version: s.Version,
			SHA256: s.SHA256, Via: s.Via, CodeID: codes[s.SkillID],
		})
	}
	a.writeJSON(w, http.StatusOK, map[string]any{
		"apiVersion":  "baton.mailloop.dev/v1alpha1",
		"kind":        "NodeSkillList",
		"items":       out,
		"next_cursor": "",
	})
}











func (a *API) handleSkillNodeCode(w http.ResponseWriter, r *http.Request) {
	p, ok := a.requireRole(w, r, console.RoleOperator)
	if !ok {
		return
	}
	skillID := r.PathValue("skill_id")
	node, err := a.resolveNode(r, r.PathValue("node_id"))
	if err != nil {
		a.failStore(w, r, err)
		return
	}
	if _, err := a.Store.Skills().Get(r.Context(), p.Identity.TenantID, skillID); err != nil {
		a.failStore(w, r, err)
		return
	}

	if r.Method == http.MethodGet {
		codes, err := a.Store.Skills().CodesForNode(r.Context(), node.NodeID)
		if err != nil {
			a.failStore(w, r, err)
			return
		}











		a.writeJSON(w, http.StatusOK, map[string]any{
			"node_id":  node.NodeID,
			"skill_id": skillID,
			"code_id":  codes[skillID],
		})
		return
	}

	var body struct {
		CodeID string `json:"code_id"`
	}
	if !a.decodeJSON(w, r, &body) {
		return
	}
	if strings.TrimSpace(body.CodeID) == "" {
		a.fail(w, r, http.StatusBadRequest, CodeInvalidRequest,
			"code_id is empty",
			"Recording an empty code would leave a row that reads as \"a code was issued\" while "+
				"naming none. A node with no code is the `no_code` state, which is the absence of "+
				"a row, not a row holding nothing.", nil)
		return
	}
	if err := a.Store.Skills().SetCode(r.Context(), node.NodeID, skillID, strings.TrimSpace(body.CodeID)); err != nil {
		a.failStore(w, r, err)
		return
	}
	a.auditSkill(r, p.Subject, "skill.code_issued", "code", skillID+" -> node:"+node.NodeID)





	a.nudgeSkills(r, []string{node.NodeID})
	w.WriteHeader(http.StatusNoContent)
}








func (a *API) handleAgentSkillArchive(w http.ResponseWriter, r *http.Request) {
	node, ok := a.requireNode(w, r)
	if !ok {
		return
	}
	id := r.PathValue("skill_id")

	assigned, err := a.Store.Skills().ForNode(r.Context(), node.TenantID, node.NodeID, node.Runtime.Template)
	if err != nil {
		a.failStore(w, r, err)
		return
	}
	var want *spi.InstalledSkill
	for i := range assigned {
		if assigned[i].SkillID == id {
			want = &assigned[i]
			break
		}
	}
	if want == nil {


		a.fail(w, r, http.StatusNotFound, CodeNotFound,
			"no such skill for this node",
			"A node can fetch only the skills installed on it or on its template.", nil)
		return
	}

	f, err := a.SkillCache.Open(want.SHA256)
	if errors.Is(err, skills.ErrNotCached) {



		if _, ferr := a.SkillCache.Fetch(r.Context(), want.SourceURL, want.SHA256); ferr != nil {
			a.fail(w, r, http.StatusBadGateway, CodeSkillFetchFailed,
				"the archive is not cached here and could not be re-fetched",
				"The cache is rebuildable from source_url + sha256. If the source is gone, "+
					"re-ingest the package with `baton skills add`.",
				map[string]any{"skill_id": id, "source_url": want.SourceURL})
			return
		}
		f, err = a.SkillCache.Open(want.SHA256)
	}
	if err != nil {
		a.fail(w, r, http.StatusInternalServerError, CodeInternal,
			"the cached archive could not be read", "", nil)
		return
	}
	defer f.Close()

	w.Header().Set("Content-Type", "application/zip")



	w.Header().Set("X-Baton-Skill-Digest", want.SHA256)
	w.WriteHeader(http.StatusOK)
	_, _ = io.Copy(w, f)
}


func (a *API) requireNode(w http.ResponseWriter, r *http.Request) (*spi.Node, bool) {
	p, ok := principalFrom(r)
	if !ok || !p.IsNode() {
		a.fail(w, r, http.StatusForbidden, CodeForbidden,
			"this endpoint requires a node certificate",
			"Nodes fetch their own skills; operators use the admin face.", nil)
		return nil, false
	}
	node, err := a.Store.Nodes().Get(r.Context(), p.Identity.Name)
	if err != nil {
		a.failStore(w, r, err)
		return nil, false
	}
	if node.Status == spi.NodeStatusRevoked {
		a.fail(w, r, http.StatusForbidden, CodeForbidden,
			"this node has been revoked", "Re-enroll with a fresh token.", nil)
		return nil, false
	}









	if node.SuspendedAt != nil {
		a.fail(w, r, http.StatusForbidden, CodeNodeSuspended,
			"this node is suspended", "An operator paused it; ask them to resume it.", nil)
		return nil, false
	}
	return node, true
}

File: cli/src/commands/skills.ts
// SPDX-License-Identifier: Apache-2.0

import { readCredentialFile } from '../credential-file.js';
import type { Client } from '../api/client.js';
import type { ParsedArgs } from '../args.js';
import { flagString } from '../args.js';
import { BatonError, ExitCode, preconditionError, usageError } from '../errors.js';
import { json, table } from '../output.js';
import { cloudFetch } from './cloud.js';
















interface SkillView {
  skill_id: string;
  name: string;
  version: string;
  sha256: string;
  source_url: string;
  size_bytes: number;
  created_at: string;
  cached: boolean;
  installs: { target_kind: string; target: string; created_at: string }[];
}








async function skillByID(client: Client, skillID: string): Promise<SkillView> {
  const list = await client.get<{ items: SkillView[] }>('/skills');
  const found = (list.items ?? []).find((s) => s.skill_id === skillID);
  if (!found) {
    throw preconditionError(
      `this control plane holds no skill ${skillID}`,
      'Run `baton skills list` to see the ids it does hold. Nothing was minted.',
    );
  }
  return found;
}


function target(args: ParsedArgs): { target_kind: string; target: string } {
  const node = flagString(args, 'node');
  const template = flagString(args, 'template');
  if (node && template) {
    throw usageError(
      'give --node or --template, not both',
      'A node install names one workspace. A template install names a scene and ' +
        'reaches nodes created from it later, including ones that do not exist yet.',
    );
  }
  if (node) return { target_kind: 'node', target: node };
  if (template) return { target_kind: 'template', target: template };
  throw usageError(
    'this needs a target',
    'Give --node <name> for one workspace, or --template <name> for every node ' +
      'set up from that template.',
  );
}

export async function skills(args: ParsedArgs, client: Client): Promise<number> {
  const sub = args.positionals[1] ?? 'list';

  if (sub === 'add') {
    const name = args.positionals[2];
    const url = flagString(args, 'url');
    const sha256 = flagString(args, 'sha256');
    if (!name || !url || !sha256) {
      throw usageError(
        'skills add needs a name, a --url and a --sha256',
        'For example: baton skills add review --url https://…/review.zip --sha256 sha256:…\n' +
          'The digest is not optional. There is no publisher signature behind a skill ' +
          'package, so it is the only check on what the source serves.',
      );
    }

    const created = await client.post<SkillView>('/skills', {
      name,
      version: flagString(args, 'version') ?? '',
      source_url: url,
      sha256,
    });

    if (args.global.output === 'json') {
      process.stdout.write(json(created) + '\n');
      return ExitCode.OK;
    }
    process.stdout.write(
      `added ${created.name} ${created.version}\n` +
        `  id      ${created.skill_id}\n` +
        `  digest  ${created.sha256}\n` +
        `  size    ${created.size_bytes} bytes\n\n` +


        `  Nothing has it yet. Install it with:\n` +
        `    baton skills install ${created.skill_id} --node <name>\n` +
        `    baton skills install ${created.skill_id} --template <name>\n`,
    );
    return ExitCode.OK;
  }
























  if (sub === 'code') {
    if (args.positionals[2] !== 'issue') {
      throw usageError(
        args.positionals[2] ? `skills code ${args.positionals[2]} is not a subcommand` : 'skills code needs a verb',
        'Today there is one: baton skills code issue --node <name> --skill <skill-id>.',
      );
    }
    const nodeRef = flagString(args, 'node');
    const skillID = flagString(args, 'skill');
    if (!nodeRef || !skillID) {
      throw usageError(
        'skills code issue needs --node and --skill',
        'A code binds one skill on one node, so both have to be named. ' +
          '`baton skills list` shows skill ids; `baton node list` shows node names.',
      );
    }































    const accountTokenPath = flagString(args, 'account-token');
    if (!accountTokenPath) {
      throw usageError(
        'skills code issue needs --account-token <path>',
        'Minting is account-gated at the registry, so this command has to present the ' +
          'account you already have. Point it at a file holding the token. Nothing was minted.',
      );
    }



    const accountToken = readCredentialFile(accountTokenPath, 'account token');


    const path = `/skills/${encodeURIComponent(skillID)}/nodes/${encodeURIComponent(nodeRef)}/code`;








    const skill = await skillByID(client, skillID);

    const seen = await client.post<{
      result: string;
      output?: string;
      data?: { content_digest?: string; code_id?: string };
    }>(`/nodes/${encodeURIComponent(nodeRef)}/console`, {
      command: 'skills.show',
      args: { skill: skill.name },
      reason: `minting a skill-code for ${skillID}`,
    });
    const digest = seen.data?.content_digest ?? '';






    if (seen.result !== 'ok' || !digest) {
      throw preconditionError(
        `${nodeRef} could not say what its copy of ${skill.name} is${seen.output ? `: ${seen.output}` : ''}`,
        'A code binds the digest of the file as that node holds it at this moment, so the ' +
          'node has to be online and holding the package. Check with `baton node show ' +
          nodeRef + '`, then try again. Nothing was minted.',
      );
    }

    const agent = flagString(args, 'agent') ?? nodeRef;
    const res = await cloudFetch('/api/dna/skill-code/issue', {
      method: 'POST',
      headers: { authorization: `Bearer ${accountToken}` },
      body: JSON.stringify({
        skill_id: skillID,
        content_digest: digest,
        agent,
        node: nodeRef,
      }),
    });
    const body = (res.body ?? {}) as Record<string, unknown>;
    if (res.status !== 200 && res.status !== 201) {
      throw new BatonError({
        code: String(body.code ?? 'SKILL_CODE_ISSUE_FAILED'),
        message: String(body.message ?? `the registry refused to mint (${res.status})`),
        remediation: String(body.remediation ?? 'Nothing was recorded on this side.'),
        exitCode: ExitCode.PRECONDITION,
      });
    }
    const codeID = String(body.code_id ?? '');
    if (!codeID) {



      throw preconditionError(
        'the registry minted a code but did not name it',
        'A code_id is what gets written into the skill and reported back at run time. ' +
          'Nothing was recorded.',
      );
    }
    await client.put(path, { code_id: codeID });

    if (args.global.output === 'json') {
      process.stdout.write(json({ skill_id: skillID, node: nodeRef, code_id: codeID }) + '\n');
      return ExitCode.OK;
    }
    process.stdout.write(
      `issued ${codeID} for ${skillID} on ${nodeRef}\n` +
        `  bound to what that node's copy says right now: ${digest}\n` +
        '  the node writes it into the skill on its next sync.\n',
    );
    return ExitCode.OK;
  }

  if (sub === 'install' || sub === 'uninstall') {
    const id = args.positionals[2];
    if (!id) {
      throw usageError(
        `skills ${sub} needs a skill id`,
        'Run `baton skills list` to see them.',
      );
    }
    const where = target(args);
    await client.post(`/skills/${encodeURIComponent(id)}/${sub}`, where);

    if (args.global.output === 'json') {
      process.stdout.write(json({ skill_id: id, ...where, action: sub }) + '\n');
      return ExitCode.OK;
    }
    const verb = sub === 'install' ? 'installed on' : 'uninstalled from';
    process.stdout.write(
      `${id} ${verb} ${where.target_kind} ${where.target}\n` +
        (where.target_kind === 'template'
          ? '  Every node set up from this template picks it up, including ones created later.\n'
          : '') +


        '  Connected nodes sync now; offline ones sync when they reconnect.\n',
    );
    return ExitCode.OK;
  }

  if (sub === 'remove') {
    const id = args.positionals[2];
    if (!id) {
      throw usageError('skills remove needs a skill id', 'Run `baton skills list` to see them.');
    }
    await client.delete(`/skills/${encodeURIComponent(id)}`);
    process.stdout.write(`removed ${id}\n  Every install of it goes too.\n`);
    return ExitCode.OK;
  }






  if (sub === 'list') {
    const body = await client.get<{ items: SkillView[] }>('/skills');
    if (args.global.output === 'json') {
      process.stdout.write(json(body) + '\n');
      return ExitCode.OK;
    }
    process.stdout.write(
      table(
        body.items,
        [
          { header: 'SKILL ID', get: (s) => s.skill_id },
          { header: 'NAME', get: (s) => s.name },
          { header: 'VERSION', get: (s) => s.version },



          { header: 'CACHED', get: (s) => (s.cached ? 'yes' : 'no') },
          {
            header: 'INSTALLED ON',
            get: (s) =>
              s.installs.length === 0
                ? '—'
                : s.installs.map((i) => `${i.target_kind}:${i.target}`).join(' '),
          },
        ],
        'no skills\n  Add one with: baton skills add <name> --url <u> --sha256 <d>\n',
      ),
    );
    return ExitCode.OK;
  }

  throw usageError(
    `unknown skills subcommand: ${sub}`,
    'Use add, list, install, uninstall, or remove. To share a skill to a network: baton resource publish --type skill.',
  );
}
